### PR TITLE
[SPARK-25253][PYSPARK] Refactor local connection & auth code

### PR DESCRIPTION
**Pull request by squito**
This eliminates some duplication in the code to connect to a server on localhost to talk directly to the jvm.  Also it gives consistent ipv6 and error handling.  Two other incidental changes, that shouldn't matter:
1) python barrier tasks perform authentication immediately (rather than waiting for the BARRIER_FUNCTION indicator)
2) for `rdd._load_from_socket`, the timeout is only increased after authentication.

### CODE REVIEW

**Comment by squito** [comment]
@jiangxb1987 @HyukjinKwon @mengxr

File: python/pyspark/java_gateway.py [hunk 1]
```diff
@@ -147,6 +147,39 @@ def do_server_auth(conn, auth_secret):
         raise Exception("Unexpected reply from iterator server.")
 
 
+def local_connect_and_auth(sock_info):
```
Comment (HyukjinKwon) [inline python/pyspark/java_gateway.py]: @squito, not a big deal but how about `local_connect_and_auth (port, auth_secret)` and ..

```python
(sockfile, sock) = local_connect_and_auth(port, auth_secret)
```

```python
(sock_file, _) = local_connect_and_auth(java_port, auth_secret)
```

```python
port, auth_secret = sock_info
(sockfile, sock) = local_connect_and_auth(port, auth_secret)
```

or 

```python
(sockfile, sock) = local_connect_and_auth(*sock_info)
```

**Review by HyukjinKwon — APPROVED** [review]

File: python/pyspark/worker.py [hunk 2]
```diff
@@ -364,8 +364,5 @@ def process():
     # Read information about how to connect back to the JVM from the environment.
     java_port = int(os.environ["PYTHON_WORKER_FACTORY_PORT"])
     auth_secret = os.environ["PYTHON_WORKER_FACTORY_SECRET"]
-    sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
-    sock.connect(("127.0.0.1", java_port))
-    sock_file = sock.makefile("rwb", 65536)
```
Comment (HyukjinKwon) [inline python/pyspark/worker.py]: @vanzin, BTW, did you test this on Windows too?

Comment (HyukjinKwon) [reply python/pyspark/worker.py]: I quickly tested and seems working fine. Please ignore this comment.

File: python/pyspark/java_gateway.py (hunk 1 above)
Comment (squito) [reply python/pyspark/java_gateway.py]: done, thanks -- I used the varargs version `*sock_info` for the last one, i forgot about that in python.

**Review by jiangxb1987 — APPROVED** [review]
LGTM, thanks!

File: python/pyspark/taskcontext.py [hunk 3]
```diff
@@ -108,38 +108,12 @@ def _load_from_socket(port, auth_secret):
...
-            continue
-        break
-    if not sock:
-        raise Exception("could not open socket")
-
-    # We don't really need a socket file here, it's just for convenience that we can reuse the
-    # do_server_auth() function and data serialization methods.
-    sockfile = sock.makefile("rwb", 65536)
-
+    (sockfile, sock) = local_connect_and_auth(port, auth_secret)
```
Comment (jiangxb1987) [inline python/pyspark/taskcontext.py]: We must set sock timeout to `None` to allow `barrier()` call blocking forever.

Comment (squito) [reply python/pyspark/taskcontext.py]: good catch, thanks!  updated

**Review by vanzin — COMMENTED** [review]
LGTM.

File: python/pyspark/java_gateway.py [hunk 4]
```diff
@@ -147,6 +147,39 @@ def do_server_auth(conn, auth_secret):
...
+        try:
+            sock.settimeout(15)
+            sock.connect(sa)
+        except socket.error as e:
+            emsg = _exception_message(e)
+            errors.append("tried to connect to %s, but an error occured: %s" % (sa, emsg))
+            sock.close()
+            sock = None
+            continue
+        break
```
[the remark this answers is not in the document]
Comment (squito) [reply python/pyspark/java_gateway.py]: done

**Comment by HyukjinKwon** [comment]
Merged to master.

**Comment by mengxr** [comment]
@squito Thanks for the refactor!

**Pull request closed without merging**